plugins/experimental/libirods-authentication_plugin-native.cpp

```diff
@@ -0,0 +1,432 @@
+                rstrcpy(md5_buf + CHALLENGE_LEN, password.c_str(), MAX_PASSWORD_LEN);
+                tty.c_lflag = oldflag;
+                if (tcsetattr(STDIN_FILENO, TCSANOW, &tty)) {
+                    printf( "Error reinstating echo mode." );
+                }
+            } // if need_password
+
+            MD5_CTX context;
+            MD5_Init( &context );
+            MD5_Update(&context, (unsigned char*)md5_buf, CHALLENGE_LEN + MAX_PASSWORD_LEN);
+
+            char digest[RESPONSE_LEN + 2];
+            MD5_Final((unsigned char*)digest, &context);
+
+            for (int i = 0; i < RESPONSE_LEN; ++i) {
+                if (digest[i] == '\0') {
+                    digest[i]++;
+                }
+            }
+
+            unsigned char out[RESPONSE_LEN*2];
+            unsigned long out_len{RESPONSE_LEN*2};
+            auto err = base64_encode((unsigned char*)digest, RESPONSE_LEN, out, &out_len);
+            if(err < 0) {
+                THROW(err, "base64 encoding of digest failed.");
+            }
+
+            resp["digest"] = std::string{(char*)out};
+            resp["next_operation"] = AUTH_CLIENT_AUTH_RESPONSE;
+
+            return resp;
+
+        } // native_auth_establish_context
+
+        json native_auth_client_request(rcComm_t* comm, const json& req)
+        {
+            json svr_req{req};
+            svr_req["next_operation"] = AUTH_AGENT_AUTH_REQUEST;
+            auto resp = request(comm, svr_req);
+
+            resp["next_operation"] = AUTH_ESTABLISH_CONTEXT;
+
+            return resp;
+
+        } // native_auth_client_request
+
+
+        #ifdef RODS_SERVER
+        json native_auth_agent_request(rsComm_t* comm, const json& req)
+        {
+            json resp{req};
+
+            char buf[ CHALLENGE_LEN + 2 ];
+            get64RandomBytes( buf );
+
+            resp["request_result"] = buf;
+
+            _rsSetAuthRequestGetChallenge(buf);
+
+            if (comm->auth_scheme) {
+                free(comm->auth_scheme);
+            }
+
+            comm->auth_scheme = strdup(AUTH_NATIVE_SCHEME.c_str());
+
+            return resp;
+
+        } // native_auth_agent_request
+        #endif
+
+
+        json native_auth_client_response(rcComm_t* comm, const json& req)
+        {
+            namespace ie = irods::experimental;
+
+            if(auto [e, n] = verify(req, {"digest","user_name","zone_name"}); !e) {
+                THROW(SYS_INVALID_INPUT_PARAM,
+                      boost::format("%s is missing %s in request")
+                      % __FUNCTION__
+                      % n);
+            }
+
+            json svr_req{req};
+            svr_req["next_operation"] = AUTH_AGENT_AUTH_RESPONSE;
+            auto resp = request(comm, svr_req);
+
+            comm->loggedIn = 1;
+
+            resp["next_operation"] = ie::AUTHENTICATION_FLOW_COMPLETE;
+
+            return resp;
+
+        } // native_auth_client_response
+
+
+        #ifdef RODS_SERVER
```

Comment:
consider keeping macro indentation separate from runtime logic. this makes it easier to spot and informs the reader that they should handle the block of code carefully.